### ISSUE TITLE
Ability to configure a smaller size cache 

### ISSUE THREAD

**Issue by huynhmb359** [coredns/coredns#2017]
The current cache capacity has the minimum cache capacity of 131,072 entries.   We did some experiment and noticed that it would consume up to 500 MB after filling up the cache.   This causes issue for running coredns on platform where memory is limited. 
Could we add support to configure cache size smaller than the current imposed minimum?

**Comment by fturib** [comment]
@huynhmb359 : what is the size you expect ?

NOTE: if you have several caches in your Corefile, please provide the size (in number of entries) you would like to be able to set up, so we have a global view of what is expected.

**Comment by huynhmb359** [comment]
@fturib,  we have several caches in our corefile.    In term of entries,  we would like to have about 5K of entries for the smallest one and 65K of entries for average.

**Comment by fturib** [comment]
Today the cache is build with 256 shards. 
Default size for each shard is 512 entries.

65k means - 256 shards of 128 entries which seems ok
5k  means - 256 shards of 20 entries which seems pretty small number of entries and will not work properly if the hashkey is not well balanced (eviction is per shard)

@miekg : what was the initial point to introduce shard ? is it concurrent access or efficiency to memory access ?

In other words, for a number as low as 20 entries, should we keep the 256 shards or switch to a lower number of shards if the max size of the cache is low ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Ability to co..." ]
>Today the cache is build with 256 shards.
>Default size for each shard is 512 entries.
>
>65k means - 256 shards of 128 entries which seems ok
>5k  means - 256 shards of 20 entries which seems pretty small number of entries and will not work properly if the hashkey is not well balanced (eviction is per shard)
>
>@miekg : what was the initial point to introduce shard ? is it concurrent access or efficiency to memory access ?

concurrency. And I think we may want to up this to 1024 by default (for even 
better concurrency)

/Miek

--
Miek Gieben

**Comment by fturib** [comment]
If we go with 1024 shards, concurrency will be better, but, what is the down side when entries per map is small ?
- memory point of view (pre-allocation of 1k maps ?)
- eviction more often (I am not sure how to compute it .. I guess that depend of the well building of the hashkey).

for information, I found here a link that show memory used by empy maps (for information): https://stackoverflow.com/questions/15313105/memory-overhead-of-maps-in-go
